### PR TITLE
Mktemp

### PR DESCRIPTION
**Pull request by KeenS**
#713 .

* quiet option is not implemented because I don't know how to control stderr
* directory option is not implemented because `TempDir` is not so flexible and `TempFile` cannot create a directory

### CODE REVIEW

File: src/mktemp/mktemp.rs [hunk 1]
```diff
@@ -0,0 +1,174 @@
+#![crate_name = "uu_mktemp"]
+
+/*
+ * This file is part of the uutils coreutils package.
+ *
+ * (c) Jordi Boggiano <j.boggiano@seld.be>
```
Comment (KeenS) [inline src/mktemp/mktemp.rs]: Is this correct? or can I use my name?

Comment (Heather) [inline src/mktemp/mktemp.rs]: you should use your name, yes

File: src/mktemp/mktemp.rs [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+
+    // >> early return options
+    let matches = match opts.parse(&args[1..]) {
+        Ok(m) => m,
+        Err(f) => crash!(1, "Invalid options\n{}", f)
+    };
+
+    if matches.opt_present("quiet") {
+        // TODO: close stderror. `crash!` macro always write output to stderro
```
Comment (ebfe) [inline src/mktemp/mktemp.rs]: If this option does nothing for now, it should not be accepted.

File: src/mktemp/mktemp.rs [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    if make_dir {
+        crash!(1, "Directory optios is not supported yet. Sorry.");
+    }
+
+
+    let tmpfile = NamedTempFileOptions::new()
+        .prefix(prefix)
+        .rand_bytes(rand)
+        .suffix(suffix)
+    // CAUTION: Creating tempfile regardless of dry_run option.
```
Comment (ebfe) [inline src/mktemp/mktemp.rs]: I don't think anything shold be created when using --dry-run, even if it is deleted later.

**Comment by ebfe** [comment]
It would be nice to squash some commits together, and/or prefix the commit message with the util name it touches.

File: src/mktemp/mktemp.rs [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    } else {
+        &matches.free[0][..]
+    };
+
+    let (prefix, rand, suffix) = match parse_template(template) {
+        Some((p, r, s)) => match suffix_opt {
+            Some(suf) => if s == "" {
+                (p, r, suf)
+            } else {
+                crash!(1, "Template should ends with 'X' when you specify suffix option.")
```
Comment (ebfe) [inline src/mktemp/mktemp.rs]: s/ends/end

File: src/mktemp/mktemp.rs [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+    let prefix = &temp[..left];
+    let rand = right - left;
+    let suffix = &temp[right..];
+    Some((prefix, rand, suffix))
+}
+
+fn exec(tmpdir: PathBuf, prefix: &str, rand: usize, suffix: &str, make_dir: bool, dry_run: bool, ) -> i32 {
+    // TODO: respect make_dir option
+    if make_dir {
+        crash!(1, "Directory optios is not supported yet. Sorry.");
```
Comment (ebfe) [inline src/mktemp/mktemp.rs]: s/optios/option

File: src/mktemp/mktemp.rs (hunk 3 above)
Comment (KeenS) [inline src/mktemp/mktemp.rs]: Yes, I agree with you. I thought you need to access file system eventually to ensure the generated random file name doesn't exist so I used this method for the sake of simplicity. However I found that `mktemp(1)` don't access the file system and don't assure the uniqueness, I'll implement dry-run correctly.

**Comment by KeenS** [comment]
All thing done.

**Comment by ebfe** [comment]
The tests fail on travis. Tests should probably check the temp file was actually created.

There should be defaults when no arguments are passed. e.g:

    # gnu works without args
    $ mktemp
    /tmp/tmp.Pv5Gvs8wsW

    $ taget/debug/uutils mktemp
    mktemp 0.0.1
    SYNOPSIS
      mktemp [OPTION]... [FILE]
    Usage:
     Create a temporary file or directory, safely, and print its name.
    TEMPLATE must contain at least 3 consecutive 'X's in last component.
    If TEMPLATE is not specified, use tmp.XXXXXXXXXX, and --tmpdir is implied
    
    Options:
        -d, --directory     Make a directory instead of a file
        -u, --dry-run       do not create anything; merely print a name (unsafe)
        -q, --quiet         Fail silently if an error occurs.
            --suffix SUFF   append SUFF to TEMPLATE; SUFF must not contain a path
                            separator. This option is implied if TEMPLATE does not
                            end with X.
        -p, --tmpdir DIR    interpret TEMPLATE relative to DIR; if DIR is not
                            specified, use $TMPDIR if set, else /tmp. With this
                            option, TEMPLATE must not be an absolute name; unlike
                            with -t, TEMPLATE may contain slashes, but mktemp
                            creates only the final component
            --help          Print this help and exit
            --version       print the version and exit

**Comment by ebfe** [comment]
Tests also fail locally for me. But just in about 1 in 10 runs. Maybe a race condition somewhere?

**Comment by KeenS** [comment]
@ebfe 

Re run without arguments: ah, I forgot to fix it

Re race condition: It is because each test case sets env var `TMPDIR`. Tests will no fail when you test it with `make test` or run with `RUST_TEST_TASKS=1`.
I'm setting TMPDIR because I don't want to leave temporary files after tests done.

Is there anyway to set `RUST_TEST_TASKS=1` only when running tests against `mktemp`? If it isn't, I'll integrate all the cases into one function so that they are run sequencially.

**Comment by ebfe** [comment]
TMPDIR should only be set in the util process not the parent test process.  There is currently no way to do that.  With #785 this can be fixed.

**Comment by KeenS** [comment]
I'll wait #785 to settle this problem.

**Comment by Heather** [comment]
@KeenS #785 merged

**Comment by KeenS** [comment]
Yeah build passing!

**Comment by ebfe** [comment]
Thanks.

**Pull request merged**